README.txt
== DESCRIPTION:

Stomp messaging server with file/dbm/memory/activerecord based FIFO
queues, queue monitoring, and basic authentication.

== SYNOPSYS:

Handles basic message queue processing

== REQUIREMENTS:

* EventMachine

== FEATURES/PROBLEMS:

=== Several queue storage backends

Handles basic message queue processing using memory, file, or dbm
based queues. Messages are sent and consumed in FIFO order (unless a
client error happens, this should be corrected in the future). Topics
are memory-only storage.  You can select activerecord, file or dbm
storage and the queues will use that, but topics will only be stored
in memory.

memory queues are of course the fastest ones but shouldn't be used if
you want to ensure all messages are delivered.

dbm queues will use berkeleydb if available, otherwise dbm or gdbm
depending on the platform. sdbm does not work well with marshalled
data. Note that these queues have not been tested in this release.

For the file based storage, each frame is stored in a single file. The
first 8 bytes contains the header length, the next 8 bytes contains
the body length, then the headers are stored as a marshalled object
followed by the body stored as a string. This storage is currently
inefficient because queues are stored separately from messages, which
forces a double write for data safety reasons on each message stored.

The activerecord based storage expects to find a database.yml file in
the configuration directory. It should be the most robust backend, but
the slowest one. The database must have an ar_messages table which can
be created with the following code (you are responsible to do so):

  ActiveRecord::Schema.define do
    create_table 'ar_messages' do |t|
      t.column 'stomp_id', :string, :null => false
      t.column 'frame', :text, :null => false
    end
  end

You can read the frames with this model:

  class ArMessage < ActiveRecord::Base
    serialize :frame
  end

The ar_message implementation will certainly change in the future.

This is meant to be easily readable by a Rails application (which
could handle the ar_messages table creation with a migration).

=== Limitations

Stompserver not support any server to server messaging (although you could
write a client to do this).

=== Monitoring

Queues can be monitored via the monitor queue (this will probably not
be supported this way in the future to avoid polluting the queue
namespace). If you subscribe to /queue/monitor, you will receive a
status message every 5 seconds that displays each queue, it's size,
frames enqueued, and frames dequeued. Stats are sent in the same
format of stomp headers, so they are easy to parse. Following is an
example of a status message containing stats for 2 queues:

Queue: /queue/client2
size: 0
dequeued: 400
enqueued: 400

Queue: /queue/test
size: 50
dequeued: 250
enqueued: 300

=== Access control

Basic client authorization is also supported.  If the -a flag is
passed to stompserver on startup, and a .passwd file exists in the run
directory, then clients will be required to provide a valid login and
passcode.  See passwd.example for the password file format.

=== Misc

Whenever you stop the server, any queues with no messages will be
removed, and the stats for that queue will be reset.  If the queue has
any messages remaining then the stats will be saved and available on
the next restart.

== INSTALL:

* gem install stompserver

stompserver will create a log, etc, and storage directory on startup
in your current working directory, the value passed to as
--working_dir parameter, or if using a config file it will
use what you specified for working_dir.  The configuration file is a
yaml file and can be specified on the command line with -C
<configfile>. A sample is provided in config/stompserver.conf.

Command line options will override options set in the yaml config
file.

To use the memory queue run as follows:
  stompserver -p 61613 -b 0.0.0.0

To use the file or dbm queue storage, use the -q switch and specificy
either file or dbm.  The file and dbm queues also take a storage
directory specified with -s.  .stompserver is the default directory if
-s is not used.
  stompserver -p 61613 -b 0.0.0.0 -q file -s .stompfile
Or
  stompserver -p 61613 -b 0.0.0.0 -q dbm -s .stompbdb

To specify where the queue is stored on disk, use the -s flag followed
by a storage directory.  To enable client authorization use -a, for
debugging use -d.
  stompserver -p 61613 -b 0.0.0.0 -q file -s .stompserver -a -d

You cannot use the same storage directory for a file and dbm queue,
they must be kept separate.

To use the activerecord queue storage use -q activerecord:
  stompserver -p 61613 -b 0.0.0.0 -q activerecord
It will try to read the etc/database.yml file in the working
directory. Here's an example of a database.yml for a PostgreSQL
database named stompserver on the 'dbserver' host usable by
PostgreSQL's user 'foo' with password 'bar'(see ActiveRecord's
documentation for the parameters needed by your database):

  adapter: postgresql
  database: stompserver
  username: foo
  password: bar
  host: dbserver

== LICENSE:

(The MIT License)

Copyright (c) 2006 Patrick Hurley
Copyright (c) 2007 Lionel Bouton

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
== 1.0.7 / 23 Aug 2013

* ACK fix, issue #1 and #2 from Doug Henderson

== 1.0.6 / 21 Apr 2011

* Change tests to use 'publish' not 'send'
* Fix broken tests under Ruby 1.9.1p180
* Handle subscriptions when 'id' header is specified

== 1.0.5 / 14 Mar 2011

* Correct exception during unsubscribe
* Replace 'hoe' with 'jeweler'

== 1.0.4 / 24 Dec 2010

* Use EM::epoll if available
* Refactor, remove unused methods
* Refactor, move recognizer and monitor to separate files
* Use hanna style docs only if hanna is present
* Smoke tested under Ruby 1.9.2p80 (2010-12-02 revision 30049) [i686-linux]

== 1.0.3 / 01 Jun 2010

* Add tested support for activerecord message stores with examples.
* Limit log message size for 'large' messages.
* Clarify some documentation.
* Add more publish / receive tests, remove duplicate tests.
* Update manifest with new file names.
* Update gemspec with new file names.

== 1.0.2 / 14 May 2010

* Continue to fix incompatabilities uncovered with gem rename.
* Add session IDs to all log messages.
* Remove some duplicate tests.

== 1.0.1 / 25 Apr 2010

* Ensure stompserver_ng and the older stompserver gems can coexist.

== 1.0.0 / 24 Apr 2010

* Rename the ruby gem.
* Start new versioning methodology.

== 0.9.9.2010.01.15.00 / 15 Jan 2010

* Refactor and correct tests.
* Correct lead sequence of config files.
* Fix send receipt sequencing bug
* Fix exceptions with no config file command line overide
* Fix problems with sysvinit script
* Change some messages to warn level

== 0.9.9.2010.01.08.00 / 08 Jan 2010

* Override server and/or port from tests.
* Fix one more 1.9 compatability issue during shutdown.
* Fix counts reported by monitor with ack => client.

== 0.9.9.2010.01.02.00 / 02 Jan 2010

* Fix server crashes when no ACK received.
* Tests for same.
* Handle SIGTERM.

== 0.9.9.2010.01.01.00 / 01 Jan 2010

* Some fixes for 'ack' => 'client' processing.  The first step down this road.
* Tests for same.  Tests are intermittently unstable due to previously unseen problems.
* Some fixes around session ID cache handling.
* Cleaner creation of the .pid file at startup.
* Sample sysvinit boot scripts - very primitive.

== 0.9.9.2009.12.27.00 / 27 Dec 2009

* Add session ID management
* Refactor stomp protocol class code

== 0.9.9.2009.12.25.00 / 25 Dec 2009

* Consolidate start messaging to a single class
* Log all runtime values
* Eliminate r19 subdirectory artifacts 
* Remove -d / --debug as the default for the test server
* Remove references to $DEBUG in the protocol handler

== 0.9.9.2009.12.21.00 / 21 Dec 2009

* Fix 1.9.0 compatability issues
* Event machine dependency is 0.12.0
* Fix mismatches between command line and configuration file parameters

== 0.9.9.2009121900 / 19 Dec 2009

* Ruby 1.9 compatability
* Use standard ruby logger instead of puts

== 0.9.9 / 31 Jan 2008

* Fix for non-memory backends with empty queues
* Fix for queue monitoring
* Use atomic checkpoint, write to a new file and then rename it into
  place to avoid incomplete qinfo files being written
* add configuration parameter for checkpoint time for better performance
  with large queues

== 0.9.8 / 16 Aug 2007

* Several storage backends instead of Madeleine (allow tradeoffs between
  robustness with ActiveRecord and speed with memory with intermediary
  solutions based on file or dbm storages).
* Better load distribution between clients acknowledging messages after
  processing them.
* Monitoring support through a dedicated queue.

== 0.9.5 / 18 Oct 2006

* Removed eventmachine dependency in the gem
  you still need eventmachine, but you can choose between
  eventmachine and eventmachine-win32, not sure how to put 
  that in the gem.
  
  Also made another minor fix to the stompserver binary.

== 0.9.4 / 17 Oct 2006

* Cleanup (wow I was tired last night)
  * Tested under Linux, added require 'rubygems' 
  * Fixed stompserver binary - it now works and takes a few simple command args
  * Fixed up minor (tesly) test syntax issues

== 0.9.3 / 16 Oct 2006

* Last one tonight I promise
  * Fixed incorrect gem requirement version on eventmachine
  * Fixed stompserver cmd added by gem

== 0.9.2 / 16 Oct 2006

* More Queue issues resolved
  * Added more unit tests
  * All tests pass
  * Queue processing done in one large batch, this may change

== 0.9.1 / 16 Oct 2006

* Stepping along
  * Fixed release issues
  * Fixed queue logic error (failed to detect offlines)

== 0.9.0 / 16 Oct 2006

* Initialial Beta Release
  * Seems to work
  * Passes numerous test cases
  * Journals using madeleine
  * Needs documentaion
  * Needs to command line processing
  * Needs service/daemon options4



lib/stomp_server_ng.rb
require 'rubygems'
require 'eventmachine'
require 'uuid'
require 'stomp_server_ng/stomp_frame'
require 'stomp_server_ng/stomp_frame_recognizer'
require 'stomp_server_ng/stomp_id'
require 'stomp_server_ng/stomp_auth'
require 'stomp_server_ng/topic_manager'
require 'stomp_server_ng/queue_manager'
require 'stomp_server_ng/qmonitor'
require 'stomp_server_ng/queue'
require 'stomp_server_ng/queue/memory_queue'
require 'stomp_server_ng/queue/file_queue'
require 'stomp_server_ng/queue/dbm_queue'
require 'stomp_server_ng/protocols/stomp'
require 'logger'

module StompServer
  VERSION = '1.0.7'
  #
  # session ID cache manager
  #
  class SessionIDManager
    #
    @@session_cache = nil
    #
    @@generator = nil
    #
    def self.initialize_cache(requested_size)
      #
      return [] if requested_size <= 0
      #
      @@session_cache = @@session_cache || []
      #
      if not @@generator
        UUID::state_file = "./.temp_uuid_state"
        @@generator = UUID::new()
      end
      #
      return @@session_cache if @@session_cache.size > requested_size
      #
      (requested_size - @@session_cache.size).times do
        @@session_cache << @@generator.generate
      end
    end
    #
    def self.get_cache_id(requested_size)
      session_id = @@session_cache.pop
      initialize_cache(requested_size) if @@session_cache.size == 0
      session_id
    end
    #
    def self.dump_cache(logger)
      logger.debug("Session ID Dump:")
      @@session_cache.each do |sessid|
        logger.debug("#{sessid}")
      end
    end
  end
  #
  # Ruby Logger Level Handler.
  #
  class LogHelper
    #
    # Set the desired logger level.
    #
    def self.set_loglevel(opts)
      @@loglevel = nil
      case opts[:log_level].downcase
        when 'debug' then @@loglevel = Logger::DEBUG
        when 'info'  then @@loglevel = Logger::INFO
        when 'warn'  then @@loglevel = Logger::WARN
        when 'error' then @@loglevel = Logger::ERROR
        else  
          @@loglevel = Logger::ERROR
      end
    end
    #
    # Return the desired logging level.
    #
    def self.get_loglevel
      @@loglevel
    end
    #
    # Display ruby version information on a defined logger output
    # destination.
    #
    def self.showversion(logger)
      # stompserver version
      logger.debug "stomp_server version: #{StompServer::VERSION}"
      # ruby version for all versions
      logger.debug "ruby: ver=#{RUBY_VERSION}p#{RUBY_PATCHLEVEL} (reldate=#{RUBY_RELEASE_DATE})"
      # more ruby version information for 1.9+
      if RUBY_VERSION =~ /1.9/
        logger.debug "ruby: rev=#{RUBY_REVISION} engine=#{RUBY_ENGINE}"
      end
    end
    #
    # Clean logging of all options values.
    #
    def self.showoptions(logger, opts)
      logger.debug("Options Display Starts")
      # This is ugly, but it should only happen once at startup ......
      if RUBY_VERSION =~ /1.8/
        opts.keys.map {|key| key.to_s}.sort.each do |str_opt|
          optname = str_opt.to_sym
          logger.debug("Option: #{optname}=#{opts[optname]}")
        end
      else  # 1.9 version
        opts.keys.sort.each do |optname|
          logger.debug("Option: #{optname}=#{opts[optname]}")
        end
      end
      logger.debug("Options Display Ends")
    end
    #
    # Show load path
    #
    def self.showloadpath(logger)
      return unless @@loglevel == Logger::DEBUG
      pos = 1
      $:.each do |lp|
        logger.debug("#{pos} #{lp}")
        pos += 1
      end
    end
  end # of class LogHelper
  #
  # Module level configuration
  #
  class Configurator

    # The final options, merged from the defaults, the config file, and the 
    # command line.
    #--
    # Should this be 'read' only after construction ????
    attr_accessor :opts

    def initialize

      @opts = nil
      @defaults = {
        #
        # For clarity maintain the same order here as below in the 'getopts' 
        # method!!!!
        #
        :auth => false,                 # -a
        :host => "127.0.0.1",           # -b
        :checkpoint => 0,               # -c
        :config => 'stompserver.conf',  # -C
        :debug => false,                # -d
        :logdir => 'log',               # -D
        :log_level => 'error',          # -l
        :logfile => 'stompserver.log',  # -L
        :port => 61613,                 # -p
        :pidfile => 'stompserver.pid',  # -P
        :queue => 'memory',             # -q
        :storage => ".stompserver",     # -s
        :session_cache => 0,            # -S
        :working_dir => Dir.getwd,      # -w
        :dbyml => 'database.yml',       # -y
        :daemon => false                # -z
      }
      # Get a crude logger
      @@log = Logger.new(STDOUT)

      # Show version numbers regardless
      StompServer::LogHelper.showversion(@@log)

      # Options handling
      @opts = getopts()   # get and merge the options

      # Finalize logger level handling
      @@log.debug "Logger Level Requested: #{@opts[:log_level].upcase}"
      StompServer::LogHelper.set_loglevel(@opts)

      # Show Load Path
      StompServer::LogHelper.showloadpath(@@log)

      @@log.level = StompServer::LogHelper.get_loglevel()

      # Turn on $DEBUG for extra debugging info if requested
      if opts[:debug]
        $DEBUG=true
        @@log.debug "-d / --debug set, $DEBUG is true"
      end

      # Configuration is complete!
      @@log.debug("#{self.class} Configuration complete")
    end

    def getopts()

      # New Options Parser
      opts_parser = OptionParser.new

      # Empty Hash for parser values
      hopts = {}

      # :auth
      opts_parser.on("-a", "--auth", String, 
        "Require client authorization") {|a| 
        hopts[:auth] = true}

      # :host
      opts_parser.on("-b", "--host=ADDR", String, 
        "Change the host (default: localhost)") {|a| 
        hopts[:host] = a}

      # :checkpoint
      opts_parser.on("-c", "--checkpoint=SECONDS", Integer, 
        "Time between checkpointing the queues in seconds (default: 0)") {|c| 
        hopts[:checkpoint] = c}

      # :config
      opts_parser.on("-C", "--config=CONFIGFILE", String, 
        "Configuration File (default: stompserver.conf)") {|c| 
        hopts[:config] = c}

      # :debug
      opts_parser.on("-d", "--debug", String, 
        "Turn on debug messages") {|d| 
        hopts[:debug] = true}

      # :logdir
      opts_parser.on("-D", "--logdir=LOGDIR", String, 
        "Log file directory  (default: log)") {|d| 
        hopts[:logdir] = d} # new

      # :log_level
      opts_parser.on("-l", "--log_level=LEVEL", String, 
        "Logger Level (default: ERROR)") {|l| 
        hopts[:log_level] = l}

      # :logfile
      opts_parser.on("-L", "--logfile=LOGFILE", String, 
        "Log file name (default: stompserver.log") {|l| 
        hopts[:logfile] = l} # new

      # :port
      opts_parser.on("-p", "--port=PORT", Integer, 
        "Change the port (default: 61613)") {|p| 
        hopts[:port] = p}

      # :pidfile
      opts_parser.on("-P", "--pidfile=PIDFILE", Integer, 
        "PID file name (default: stompserver.pid)") {|p| 
        hopts[:pidfile] = p} # new

      # :queue
      opts_parser.on("-q", "--queuetype=QUEUETYPE", String, 
        "Queue type (memory|dbm|activerecord|file) (default: memory)") {|q| 
        hopts[:queue] = q}

      # :storage
      opts_parser.on("-s", "--storage=DIR", String, 
        "Change the storage directory (default: .stompserver, relative to working_dir)") {|s| 
        hopts[:storage] = s}

      # :session_cache
      opts_parser.on("-S", "--session_cache=SIZE", Integer, 
        "session ID cache size (default: 0, disable session ID cache)") {|s| 
        hopts[:session_cache] = s}

      # :working_dir
      opts_parser.on("-w", "--working_dir=DIR", String, 
        "Change the working directory (default: current directory)") {|s| 
        hopts[:working_dir] = s}

      # :dbyml
      opts_parser.on("-y", "--dbyml=YMLFILE", String, 
        "Database .yml file name (default: database.yml)") {|y| 
        hopts[:dbyml] = y}

      # :daemon
      opts_parser.on("-z", "--daemon", String, 
        "Daemonize server process") {|d| 
        hopts[:daemon] = true}

      # Handle help if required
      opts_parser.on("-h", "--help", "Show this message") do
        puts opts_parser
        exit
      end

      opts_parser.parse(ARGV)

      # Handle the config file
      config_found = false
      # Load a default config file first if it exists.
      loaded_opts = {}
      full_path_default = File.expand_path(@defaults[:config])
      if File.exists?(full_path_default)
        @@log.debug("Loading config file defaults from: #{full_path_default}")
        loaded_opts.merge!(YAML.load_file(full_path_default))
        @defaults[:config] = full_path_default
        config_found = true
      end
      # If a command line specified config file exists, overlay any new
      # parameters in that file.
      if hopts[:config]
        full_path_cl = File.expand_path(hopts[:config])
        if File.exists?(full_path_cl)
          @@log.debug("Loading config file overrides from: #{full_path_cl}")
          loaded_opts.merge!(YAML.load_file(full_path_cl))
          hopts[:config] = full_path_cl
          config_found = true
        end
      end
      @@log.warn("No configuration file found.") unless config_found

      # Run basic required merges on all the options
      opts = {}                         # set to empty
      opts = opts.merge(@defaults)      # 01 = merge in defaults
      opts = opts.merge(loaded_opts)    # 02 = merge in loaded from config file
      opts = opts.merge(hopts)          # 03 = merge in command line options

      # Last but not least: Miscellaneous file definitions
      opts[:etcdir] = File.join(opts[:working_dir],'etc')           # Define ':etcdir'
      opts[:storage] = File.join(opts[:working_dir],opts[:storage]) # Override! ':storage'
      opts[:logdir] = File.join(opts[:working_dir],opts[:logdir])   # Override! ':logdir'
      opts[:logfile] = File.join(opts[:logdir],opts[:logfile])      # Override! ':logfile'
      opts[:pidfile] = File.join(opts[:logdir],opts[:pidfile])      # Override! ':pidfile'
      # :dbyml will be a full path and file name
      unless File.exists?(File.expand_path(opts[:dbyml]))
        opts[:dbyml] = File.join(opts[:etcdir],opts[:dbyml])        # Override! ':dbyml'
      else
        opts[:dbyml] = File.expand_path(opts[:dbyml])
      end

      # Authorization - working file
      if opts[:auth]
        opts[:passwd] = File.join(opts[:etcdir],'.passwd')
      end
      
      # Return merged values (in Hash)
      return opts
    end
  end

  #
  # Run server start up.
  #
  class Run
    #
    attr_accessor :queue_manager, :auth_required, :stompauth, :topic_manager

    #
    attr_accessor :session_cache

    # Intiialize
    def initialize(opts)
      @@log = Logger.new(STDOUT)
      @opts = opts
#      @@log = Logger.new(@opts[:logfile])
#      StompServer::LogHelper.set_loglevel(@opts)
      @@log.level = StompServer::LogHelper.get_loglevel()

      @queue_manager = nil
      @auth_required = nil
      @stompauth = nil
      @topic_manager = nil
      @session_cache = nil
      @@log.debug("#{self.class} Run class initialize method complete")
    end

    # Server stop on SIGINT or SIGTERM
    def stop(pidfile)
      @queue_manager.stop("KILL_ISSUED")
      @@log.debug "Stompserver #{StompServer::VERSION} shutting down"
      STDOUT.flush
      EventMachine::stop_event_loop
      File.delete(pidfile)
    end

    # Startup
    def start
      @@log.debug("#{self.class}.start begins")

      # Handle group priviliges!
      # N.B.: Handle these options from the command line ?????????
      begin
        if @opts[:group]
          @@log.debug "Changing group to #{@opts[:group]}."
          Process::GID.change_privilege(Etc.getgrnam(@opts[:group]).gid)
        end

        if @opts[:user]
          @@log.debug "Changing user to #{@opts[:user]}."
          Process::UID.change_privilege(Etc.getpwnam(@opts[:user]).uid)
        end
      rescue Errno::EPERM
        @@log.error "FAILED to change user:group #{@opts[:user]}:#{@opts[:group]}: #$!"
        exit 1
      end

      # Make required directories unless they already exist
      Dir.mkdir(@opts[:working_dir]) unless File.directory?(@opts[:working_dir])
      Dir.mkdir(@opts[:logdir]) unless File.directory?(@opts[:logdir])
      Dir.mkdir(@opts[:etcdir]) unless File.directory?(@opts[:etcdir])

      # Determine qstore type
      if @opts[:queue] == 'dbm'
        qstore=StompServer::DBMQueue.new(@opts[:storage])
        @@log.debug "Queue storage is DBM"
      elsif @opts[:queue] == 'file'
        qstore=StompServer::FileQueue.new(@opts[:storage])
        @@log.debug "Queue storage is FILE"
      elsif @opts[:queue] == 'activerecord'
        require 'stomp_server_ng/queue/activerecord_queue'
        qstore=StompServer::ActiveRecordQueue.new(@opts[:etcdir], @opts[:storage], @opts[:dbyml])
        @@log.debug "Queue storage is ActiveRecord"
      else
        qstore=StompServer::MemoryQueue.new
        @@log.debug "Queue storage is MEMORY"
      end

      # Set checkpoint interval
      qstore.checkpoint_interval = @opts[:checkpoint]
      @@log.debug "Checkpoint interval is #{qstore.checkpoint_interval}" if $DEBUG

      #
      @topic_manager = StompServer::TopicManager.new
      @queue_manager = StompServer::QueueManager.new(qstore)
      @@log.debug("Managers are initialized.")
      @@log.debug("Topic Manager: #{@topic_manager}")
      @@log.debug("Queue Manager: #{@queue_manager}")

      # Authorization: requirement
      @auth_required = @opts[:auth]
      if @auth_required
        @stompauth = StompServer::StompAuth.new(@opts[:passwd])
      end

      # Initialize session ID cache
      if @opts[:session_cache] > 0
        StompServer::SessionIDManager.initialize_cache(@opts[:session_cache])
        StompServer::SessionIDManager.dump_cache(@@log);
      end

      # If we are going to daemonize, it should be almost the last
      # thing we do here.
      @@log.debug("#{self.class}.start Daemonize: #{@opts[:daemon]}")
      if @opts[:daemon]
        @@log.debug("#{self.class}.start going to background")
        @@log.debug("#{self.class}.start check #{@opts[:logfile]}")

        StompServer::LogHelper.showversion(@@log)

        STDOUT.flush    # clear the decks
        Daemonize.daemonize(log_file=@opts[:logfile])
        # change back to the original starting directory
        Dir.chdir(@opts[:working_dir])
      end

      # But write pidfile only after possible daemonization
      curr_pid = Process.pid
      open(@opts[:pidfile],"w") {|f| f.write(curr_pid) }
      @@log.debug("Pid File Contents: #{curr_pid}")

      # OK, log and set the SIGINT signal handler.
      @@log.debug("#{self.class}.start setting trap at completion")
      StompServer::LogHelper.showversion(@@log) # one more time at startup
      # Show Load Path
      StompServer::LogHelper.showloadpath(@@log)
      StompServer::LogHelper.showoptions(@@log, @opts) # Dump runtime options
      trap("INT") { @@log.debug "INT signal received.";stop(@opts[:pidfile]) }
      trap("TERM") { @@log.debug "TERM signal received.";stop(@opts[:pidfile]) }
    end
  end # of class Run
#
end # of module StompServer

